src/Microsoft.Health.Fhir.SqlServer/Features/Schema/Sql/Sprocs/GetResourceVersions.sql

```diff
@@ -13,15 +13,24 @@ BEGIN TRY
   SELECT A.ResourceTypeId
         ,A.ResourceId
         ,A.ResourceSurrogateId
-        -- set version to 0 if there is no gap available. It would indicate conflict for the caller.
-        ,Version = CASE 
-                     WHEN EXISTS (SELECT * FROM dbo.Resource B WHERE B.ResourceTypeId = A.ResourceTypeId AND B.ResourceSurrogateId = A.ResourceSurrogateId) THEN 0
-                     WHEN isnull(U.Version, 1) - isnull(L.Version, 0) > 1 THEN isnull(U.Version, 1) - 1 
-                     ELSE 0 
+        -- set version to 0 if there is no gap available, or lastUpdated is already used. It would indicate conflict for the caller.
+        ,Version = CASE
+                     -- ResourceSurrogateId is generated from lastUpdated only without extra bits at the end. Need to ckeck interval (0..79999) on resource id level.
+                     WHEN EXISTS (SELECT * 
+                                    FROM dbo.Resource B WITH (INDEX = IX_Resource_ResourceTypeId_ResourceId_Version)
+                                    WHERE B.ResourceTypeId = A.ResourceTypeId 
+                                      AND B.ResourceId = A.ResourceId 
+                                      AND B.ResourceSurrogateId BETWEEN A.ResourceSurrogateId AND A.ResourceSurrogateId + 79999
```

Comment:
Why 79999? Is there significance to this number?

Reply:
Sure there is. This is baked into surrogate id definition. Look at ResourceSurrogateIdUniquifierSequence